File: backend/product-service/src/main/java/com/ecommerce/product/service/ProductService.java
package com.ecommerce.product.service;
import com.ecommerce.product.dto.response.ProductWithInventoryResponse;
import com.ecommerce.product.dto.request.ProductRequest;
import com.ecommerce.product.dto.response.ProductResponse;
import com.ecommerce.product.dto.response.ProductWithInventoryResponse;
import com.ecommerce.product.enums.ProductCategory;
import com.ecommerce.product.enums.ProductStatus;
import org.springframework.data.domain.Page;

public interface ProductService {

    ProductResponse createProduct(ProductRequest request);

    ProductResponse updateProduct(Long id, ProductRequest request);

    ProductResponse getProductById(Long id);

    ProductResponse getProductBySku(String sku);

    Page<ProductResponse> getAllProducts(
            int page,
            int size,
            String sortBy
    );

    Page<ProductResponse> getProductsByCategory(
            ProductCategory category,
            int page,
            int size,
            String sortBy
    );

    Page<ProductResponse> getProductsByBrand(
            String brand,
            int page,
            int size,
            String sortBy
    );

    Page<ProductResponse> getProductsByStatus(
            ProductStatus status,
            int page,
            int size,
            String sortBy
    );

    void deleteProduct(Long id);

    ProductWithInventoryResponse
    getProductWithInventory(Long id);


}

File: frontend/src/services/productService.js
import api from './api'

export const getProducts = async ({
                                      page = 0,
                                      size = 10,
                                      sortBy = 'id',
                                  } = {}) => {
    const response = await api.get('/api/v1/products', {
        params: {
            page,
            size,
            sortBy,
        },
    })

    return response.data
}


export const createProduct = async (productData) => {
    const response = await api.post(
        '/api/v1/products',
        productData
    )

    return response.data
}


export const updateProduct = async (id, productData) => {
    const response = await api.put(
        `/api/v1/products/${id}`,
        productData
    )

    return response.data
}


export const deleteProduct = async (id) => {
    const response = await api.delete(
        `/api/v1/products/${id}`
    )

    return response.data
}


export const getProductDetails = async (productId) => {
    const response = await api.get(
        `/api/v1/products/${productId}/details`
    )

    return response.data
}